silverorange/sogulp
# Update gulp to v4

This upgrades the sogulp package from the gulp 3.x API to the gulp 4.x API. Gulp v4 is required for Node 14 upgrade:

- defined tasks are node exports instead of using gulp.task
- series/parallel API used for composition of task dependencies
- remove deprecated gulp-utils package
- use gulp.watch instead of chokidar directly for fs watching
- cleanups and bug fixes
- convert tasks to async functions where appropriate
- apply prettier to project
- commit yarn.lock for development
- update eslint to modern version

Previously https://github.com/silverorange/sogulp/pull/28/

See video for more context https://www.dropbox.com/s/f3lrp8ar8s9ymb6/Gulp-sogulp-Version-4-2020-10-09.mp4?dl=0

Testing
=======
1. On berna or roble
2. Check out this pr in the `/so/packages/sogulp/work-$(whoami)` folder
3. Go to a PHP site like course-host (berna) or pcrap (roble)
4. Run `yarn remove gulp sogulp && yarn add --dev gulp@4.0.2 file:/so/packages/sogulp/work-$(whoami)`
5. Remove the old gulpfile symlink `rm gulpfile.js`
6. Add a new `gulpfile.js` file that contains:
```js
const tasks = require('sogulp');
module.exports = tasks;
```
7. See if things are working with `./node_modules/.bin/gulp --tasks`. This should list registered tasks.
8. Run some tasks `./node_modules/.bin/gulp concentrate`
9. Try using the `--symlinks` flag: `./node_modules/.bin/gulp --symlinks=swat,site,store concentrate`
10. Try using the default watch task: `./node_modules/.bin/gulp --symlinks=swat,site,store`
11. While watcher is running, try:
  - create new php file
  - edit existing php file
  - delete existing php file
  - add syntax error to a php file
  - add new less file
  - edit existing less file
  - delete existing less file
12. Use ctrl+c to stop watching
13. Verify symlinks are restored in `vendor/silverorange/`
